### PR TITLE
Allow devices to specify if they want to disregard grouping

### PR DESCRIPTION
**Pull request by kshyatt-aws**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

Add variable in `execute` to allow devices to request that Hamiltonians **not** be expanded using grouped indices, even if they are known.

**Description of the Change:**

Default behaviour is unchanged. If a device does have a `use_grouping` attr, it can specify that grouping not be used.

**Benefits:**

Allows devices to control whether one expectation value or many are calculated. Avoids unexpected conversion of one observable (a Hamiltonian) into a list of observables.

**Possible Drawbacks:**

Degraded performance if device developers choose badly when setting the default for their device.

**Related GitHub Issues:**

### CODE REVIEW

**Review by albi3ro — COMMENTED** [review]
Thanks for the PR.

While we have default behaviour with the `getattr`, we should also define default behaviour by updating `_capabilites` to have `"use_grouping": True` on line 128 of `_device.py`.  This will keep track of what keys are possible.

Do you want me to write a unittest for this flag?

**Comment by kshyatt-aws** [comment]
> Do you want me to write a unittest for this flag?

Yes please. My Python monkeypatching knowledge is minimal at best.

**Comment by albi3ro** [comment]
I thought adding it in to the default configuration would be painless. Looks like it is not. Feel free to remove it from the default configuration dictionary if it breaks too many tests.

If this were a long-term solution, I would want it in the configuration dictionary. But we will be moving away from `Device` soon enough that I'm not as concerned about the effect of technical debt.

File: pennylane/_device.py [hunk 1]
```diff
@@ -733,16 +733,16 @@ def batch_transform(self, circuit):
             for obs in circuit.observables
             if obs.name == "Hamiltonian"
         )
+        use_grouping = getattr(self, "use_grouping", True)
```
Comment (albi3ro) [inline pennylane/_device.py:736]:
```suggestion
        # device property present in braket plugin
        use_grouping = getattr(self, "use_grouping", True)
```

**Comment by kshyatt-aws** [comment]
Thank you @albi3ro for the test.

File: pennylane/_device.py [hunk 2]
```diff
@@ -733,16 +733,17 @@ def batch_transform(self, circuit):
...
         )
+        # device property present in braket plugin
+        use_grouping = getattr(self, "use_grouping", True)
 
         hamiltonian_in_obs = "Hamiltonian" in [obs.name for obs in circuit.observables]
 
         return_types = [m.return_type for m in circuit.observables]
 
         is_shadow = ShadowExpval in return_types
+        hamiltonian_usable = not supports_hamiltonian or (finite_shots and not is_shadow)
```
Comment (albi3ro) [inline pennylane/_device.py:744]: Maybe `hamiltonain_unusable`? If the device doesn't support hamiltonians, then the hamiltonian will *not* be usable by the device.

File: tests/test_device.py [hunk 3]
```diff
@@ -1022,3 +1022,37 @@ def test_result_empty_tape(self, mock_device_with_paulis_and_methods, tol):
         assert np.allclose(
             res[0], dev.execute(empty_tape.operations, empty_tape.observables), rtol=tol, atol=0
         )
+
+class TestGrouping:
+    """Tests for the use_grouping option for devices."""
+    @pytest.mark.parametrize("use_grouping", (True, False))
+    def test_batch_transform_checks_use_grouping_property(use_grouping):
```
Comment (albi3ro) [inline tests/test_device.py:1029]:
```suggestion
    def test_batch_transform_checks_use_grouping_property(self, use_grouping):
```
If the test is in a class.

File: pennylane/_device.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/_device.py:744]: @kshyatt-aws I'd like to clarify the name of the explanatory variable, but then I'm good to approve.

**Pull request merged**